### PR TITLE
[staging] coreutils: 9.0 -> 9.1

### LINKED ISSUE #168901: coreutils 9.1
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: coreutils

current version: 9.0
desired version: 9.1

###### Notify maintainers

maintainers: @dasJ 

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by dasJ**
###### Description of changes

Closes #168901

###### Things done

- Built on platform(s)
  - [x] x86_64-linux → built stdenv
  - [ ] aarch64-linux
  - [ ] x86_64-darwin → still building stdenv
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
@GrahamcOfBorg build coreutils

**Review by alyssais — APPROVED** [review]
I tested some scripts that exercise a fair bit of coreutils on x86_64-linux.  I also verified that cat does indeed use copy_file_range, as promised, which was a nice speedup for my scripts that make big disk images. :)

**Review by ivan — APPROVED** [review]
I applied this to an x86_64 NixOS master and rebuilt and booted into the system and did not encounter issues. I tested `cp -a` and `cp -a --reflink=never` on btrfs and did not see issues. I looked through https://lists.gnu.org/archive/html/bug-coreutils/ and did not see any 9.1-specific reports.

**Comment by dasJ** [comment]
@edolstra would you be okay with being removed from the maintainers to reduce your inbox spam? Or do you still want to get notified on coreutils-related PRs?

**Comment by dasJ** [comment]
I think this is good to go now

**Comment by vcunat** [comment]
What about another commit that replaces the `optionalAttrs` by a normal list with optional parts?  (Though it's not related to the update, except for also causing a rebuild I think.)

**Comment by dasJ** [comment]
Do you mean the musl and android things?

File: pkgs/tools/misc/coreutils/default.nix [hunk 1]
```diff
@@ -117,6 +117,10 @@ stdenv.mkDerivation (rec {
 
   NIX_LDFLAGS = optionalString selinuxSupport "-lsepol";
   FORCE_UNSAFE_CONFIGURE = optionalString stdenv.hostPlatform.isSunOS "1";
+  NIX_CFLAGS_COMPILE =
+    # Work around a bogus warning in conjunction with musl.
+    (optionalString stdenv.hostPlatform.isMusl "-Wno-error")
+    + (optionalString stdenv.hostPlatform.isAndroid "-D__USE_FORTIFY_LEVEL=0");
```
Comment (vcunat) [inline pkgs/tools/misc/coreutils/default.nix:123]: Uh well, maybe musl on android doesn't make sense, but perhaps at least put an extra space in there?

Generally I tend to use lists, as those behave nicer:
```nix
optional stdenv.hostPlatform.isMusl "-Wno-error"
++ optional stdenv.hostPlatform.isAndroid "-D__USE_FORTIFY_LEVEL=0"
```
but I don't mind either way.

**Comment by dasJ** [comment]
Seems unrelated, I'm touching coreutils and not pkg-config :thinking:

**Comment by risicle** [comment]
I suspect this of causing my macos 10.15 system to produce garbage binaries in some circumstances. Reproducible case `${ffmpeg}/lib/libavcodec.dylib`. It's definitely corrupted by something *late* in ffmpeg's build, after `checkPhase` but before `fixupPhase`. Might be a matter of figuring out exactly what the `installPhase` is doing.

**Comment by risicle** [comment]
It looks a lot like a resurgence of https://debbugs.gnu.org/cgi/bugreport.cgi?bug=51433 to me. Might we be able to reintroduce the patch just for darwin-x86_64?

**Comment by ivan** [comment]
Why is there a resurgence on your macOS? Are you using ZFS or another filesystem? If ZFS, is there a fixed version you can use?

**Comment by risicle** [comment]
> Why is there a resurgence on your macOS?

Well I presume because the patch was removed. Am currently building with the patch restored to check that fixes it.

> Are you using ZFS or another filesystem?

Nope, using apfs.

**Pull request merged**